In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
data = pd.read_excel('/Users/xuetong/Downloads/International Data Analyst Take Home Excercise 2.xlsx',sheet_name='Data')
data.head(10)

Date Customer Type     Delivery Type Device Type  \
0 2017-09-01      Existing  Store Collection    Desktop    
1 2017-09-01           New     Home Delivery    Desktop    
2 2017-09-01           New  Store Collection    Desktop    
3 2017-09-01           New     Home Delivery     Mobile    
4 2017-09-01      Existing  Store Collection     Mobile    
5 2017-09-01      Existing     Home Delivery     Mobile    
6 2017-09-01      Existing     Home Delivery    Desktop    
7 2017-09-01           New  Store Collection     Mobile    
8 2017-09-02      Existing  Store Collection     Mobile    
9 2017-09-02      Existing     Home Delivery    Desktop    

   Before You Go Unique Daily Users  Before You Go Cart Adds   \
0                              4910                      1929   
1                              4450                      1753   
2                               487                       148   
3                               398                       103   
4                              5159                      1750   
5                             20008                      7045   
6                             20647                      7762   
7                                59                        15   
8                              5022                      1788   
9                             20362                      8168   

   Before You Go Total Revenue From Cart Adds  Book Slot Unique Daily Users  \
0                                        3737                          4839   
1                                        2470                          4166   
2                                         257                           478   
3                                         153                           343   
4                                        3356                          4785   
5                                       13190                         18655   
6                                       15198                         19739   
7                                          20                            52   
8                                        3194                          4828   
9                                       15373                         19382   

   Select Payment Unique Daily Users  Order Summary Unique Daily Users  \
0                               4342                              4245   
1                               3608                              3504   
2                                452                               435   
3                                272                               261   
4                               4242                              4059   
5                              18014                             17221   
6                              19157                             18058   
7                                 50                                46   
8                               4058                              3854   
9                              18631                             18446   

   Orders    Revenue  
0    4037   379289.0  
1    3450   311501.0  
2     399    36074.0  
3     250    17452.0  
4    3855   360265.0  
5   17027  1608022.0  
6   17673  1652713.0  
7      43     3112.0  
8    3800   350621.0  
9   18023  1711522.0

In [61]:
data.columns

Index(['Date', 'Customer Type', 'Delivery Type', 'Device Type',
       'Before You Go Unique Daily Users', 'Before You Go Cart Adds ',
       'Before You Go Total Revenue From Cart Adds',
       'Book Slot Unique Daily Users', 'Select Payment Unique Daily Users',
       'Order Summary Unique Daily Users', 'Orders', 'Revenue',
       'BFYG_to_BS_CR', 'BS_to_payment_CR', 'payment_to_summary_CR',
       'summary_to_order_CR', 'recommendation_acceptance_rate',
       'recommendation_revenue_per_user', 'revenue_per_order'],
      dtype='object')

## Converstion Rate Analysis

In [3]:
data['BFYG_to_BS_CR'] = data.apply(lambda r: r['Book Slot Unique Daily Users']/r['Before You Go Unique Daily Users']*100,axis = 1).round(1)
data['BS_to_payment_CR'] = data.apply(lambda r: r['Select Payment Unique Daily Users']/r['Book Slot Unique Daily Users']*100, axis = 1).round(1)
data['payment_to_summary_CR'] = data.apply(lambda r: r['Order Summary Unique Daily Users']/r['Select Payment Unique Daily Users']*100, axis = 1).round(1)
data['summary_to_order_CR'] = data.apply(lambda r: r['Orders']/r['Order Summary Unique Daily Users']*100, axis = 1).round(1)

In [4]:
data.head()

Date Customer Type     Delivery Type Device Type  \
0 2017-09-01      Existing  Store Collection    Desktop    
1 2017-09-01           New     Home Delivery    Desktop    
2 2017-09-01           New  Store Collection    Desktop    
3 2017-09-01           New     Home Delivery     Mobile    
4 2017-09-01      Existing  Store Collection     Mobile    

   Before You Go Unique Daily Users  Before You Go Cart Adds   \
0                              4910                      1929   
1                              4450                      1753   
2                               487                       148   
3                               398                       103   
4                              5159                      1750   

   Before You Go Total Revenue From Cart Adds  Book Slot Unique Daily Users  \
0                                        3737                          4839   
1                                        2470                          4166   
2                                         257                           478   
3                                         153                           343   
4                                        3356                          4785   

   Select Payment Unique Daily Users  Order Summary Unique Daily Users  \
0                               4342                              4245   
1                               3608                              3504   
2                                452                               435   
3                                272                               261   
4                               4242                              4059   

   Orders   Revenue  BFYG_to_BS_CR  BS_to_payment_CR  payment_to_summary_CR  \
0    4037  379289.0           98.6              89.7                   97.8   
1    3450  311501.0           93.6              86.6                   97.1   
2     399   36074.0           98.2              94.6                   96.2   
3     250   17452.0           86.2              79.3                   96.0   
4    3855  360265.0           92.8              88.7                   95.7   

   summary_to_order_CR  
0                 95.1  
1                 98.5  
2                 91.7  
3                 95.8  
4                 95.0

In [36]:
BFYG_to_BS_CR_median = data['BFYG_to_BS_CR'].median()
BS_to_payment_CR_median = data['BS_to_payment_CR'].median()
payment_to_summary_CR_median = data['payment_to_summary_CR'].median()
summary_to_order_CR_median = data['summary_to_order_CR'].median()

In [37]:
array = [BFYG_to_BS_CR_median,BS_to_payment_CR_median,payment_to_summary_CR_median,summary_to_order_CR_median]

In [38]:
pd_array = pd.DataFrame(array, index = ['beforeyougo_to_bookslot','bookslot_to_payment','payment_to_summary','summary_to_order'],
                       columns = ['conversion_rate_median'])
## conversion rate of before you go to book slot and book slot to select payment is low.

In [40]:
pd_array

conversion_rate_median
beforeyougo_to_bookslot                   90.15
bookslot_to_payment                       92.20
payment_to_summary                        96.30
summary_to_order                          96.60

In [10]:
### 
CR = ['BFYG_to_BS_CR','BS_to_payment_CR','payment_to_summary_CR','summary_to_order_CR']
CR_customer_type = data.groupby(['Customer Type'])['BFYG_to_BS_CR','BS_to_payment_CR','payment_to_summary_CR','summary_to_order_CR'].median()

In [11]:
CR_customer_type
### new customer converstion rate is lower than existing customer convertion rate

BFYG_to_BS_CR  BS_to_payment_CR  payment_to_summary_CR  \
Customer Type                                                           
Existing               93.15              93.2                   97.3   
New                    87.85              89.9                   95.4   

               summary_to_order_CR  
Customer Type                       
Existing                     97.90  
New                          94.75

In [12]:
CR_Delivery_Type = data.groupby(['Delivery Type'])['BFYG_to_BS_CR','BS_to_payment_CR','payment_to_summary_CR','summary_to_order_CR'].median()
CR_Delivery_Type

BFYG_to_BS_CR  BS_to_payment_CR  payment_to_summary_CR  \
Delivery Type                                                              
Home Delivery              92.0             89.40                   96.2   
Store Collection           89.2             93.05                   96.6   

                  summary_to_order_CR  
Delivery Type                          
Home Delivery                   97.60  
Store Collection                95.15

In [13]:
CR_device_type = data.groupby(['Device Type'])['BFYG_to_BS_CR','BS_to_payment_CR','payment_to_summary_CR','summary_to_order_CR'].median()
CR_device_type
### mobile convertion rate is lower than Desktop convertion rate

BFYG_to_BS_CR  BS_to_payment_CR  payment_to_summary_CR  \
Device Type                                                           
Desktop              93.70              92.6                  97.15   
Mobile               87.65              91.8                  95.65   

             summary_to_order_CR  
Device Type                       
Desktop                    97.25  
Mobile                     95.60

In [14]:
CR_all_group = data.groupby(['Customer Type', 'Delivery Type', 'Device Type'])['BFYG_to_BS_CR','BS_to_payment_CR','payment_to_summary_CR','summary_to_order_CR'].median()
CR_all_group 
###  new + home delivery + mobile  convertion rate of 

BFYG_to_BS_CR  BS_to_payment_CR  \
Customer Type Delivery Type    Device Type                                    
Existing      Home Delivery    Desktop              95.05             96.20   
                               Mobile               94.75             92.30   
              Store Collection Desktop              92.75             93.60   
                               Mobile               91.45             91.55   
New           Home Delivery    Desktop              93.65             86.95   
                               Mobile               85.80             79.80   
              Store Collection Desktop              93.95             92.00   
                               Mobile               85.35             95.60   

                                            payment_to_summary_CR  \
Customer Type Delivery Type    Device Type                          
Existing      Home Delivery    Desktop                      95.90   
                               Mobile                       98.05   
              Store Collection Desktop                      97.25   
                               Mobile                       97.25   
New           Home Delivery    Desktop                      97.15   
                               Mobile                       92.95   
              Store Collection Desktop                      97.45   
                               Mobile                       93.85   

                                            summary_to_order_CR  
Customer Type Delivery Type    Device Type                       
Existing      Home Delivery    Desktop                    98.20  
                               Mobile                     98.75  
              Store Collection Desktop                    97.75  
                               Mobile                     96.70  
New           Home Delivery    Desktop                    97.05  
                               Mobile                     95.60  
              Store Collection Desktop                    94.80  
                               Mobile                     91.50

## recommendation analysis 

In [63]:
### recommendation efficiency
data['recommendation_acceptance_rate'] = data.apply(lambda r: r['Before You Go Cart Adds ']/r['Before You Go Unique Daily Users']*100,axis = 1)
data['recommendation_revenue_per_user'] = data.apply(lambda r: r['Before You Go Total Revenue From Cart Adds']/r['Before You Go Cart Adds '],axis =1)

In [64]:
recommendation_acceptance_rate_median = data['recommendation_acceptance_rate'].median()
recommendation_revenue_per_user_median = data['recommendation_revenue_per_user'].median()

In [65]:
recommendation_acceptance_rate_median
### recommendation efficiency could be improved

36.594055391002044

In [66]:
recommendation_revenue_per_user_median

1.7743893906110162

## Revenue Analysis 

In [43]:
data['revenue_per_order'] = data.apply(lambda r:r['Revenue']/r['Orders'], axis = 1)

In [45]:
revnue_per_order_customer_type = data.groupby(['Customer Type'])['revenue_per_order'].median()
revnue_per_order_customer_type

Customer Type
Existing    93.319287
New         80.546326
Name: revenue_per_order, dtype: float64

In [46]:
revenue_delievery_type  = data.groupby(['Delivery Type'])['revenue_per_order'].median()
revenue_delievery_type

Delivery Type
Home Delivery       91.682106
Store Collection    87.935114
Name: revenue_per_order, dtype: float64

In [47]:
revenue_device_type = data.groupby(['Device Type'])['revenue_per_order'].median()
revenue_device_type

Device Type
Desktop     90.876481
Mobile      81.537730
Name: revenue_per_order, dtype: float64

In [48]:
revenue = data.groupby(['Customer Type', 'Delivery Type', 'Device Type'])['revenue_per_order'].median()
revenue

Customer Type  Delivery Type     Device Type
Existing       Home Delivery     Desktop        97.473773
                                 Mobile         94.408730
               Store Collection  Desktop        90.861119
                                 Mobile         89.840027
New            Home Delivery     Desktop        89.559456
                                 Mobile         74.219123
               Store Collection  Desktop        87.829068
                                 Mobile         71.651163
Name: revenue_per_order, dtype: float64

## Additional Analysis

In [56]:
order_by_customer_type = data.groupby(['Customer Type'])['Orders'].sum()
order_by_customer_type

Customer Type
Existing    1237470
New          120010
Name: Orders, dtype: int64

In [57]:
Orders_by_delievery_type  = data.groupby(['Delivery Type'])['Orders'].sum()
Orders_by_delievery_type

Delivery Type
Home Delivery       1110641
Store Collection     246839
Name: Orders, dtype: int64

In [58]:
Orders_by_device_type = data.groupby(['Device Type'])['Orders'].sum()
Orders_by_device_type

Device Type
Desktop     742486
Mobile      614994
Name: Orders, dtype: int64

In [59]:
orders = data.groupby(['Customer Type', 'Delivery Type', 'Device Type'])['Orders'].sum()
orders

Customer Type  Delivery Type     Device Type
Existing       Home Delivery     Desktop        510839
                                 Mobile         492717
               Store Collection  Desktop        119990
                                 Mobile         113924
New            Home Delivery     Desktop         99935
                                 Mobile           7150
               Store Collection  Desktop         11722
                                 Mobile           1203
Name: Orders, dtype: int64

## Summary

1. The conversion rate of first two step is low.

In [33]:
pd_array

conversion_rate_median
beforeyougo_to_bookslot                   90.15
bookslot_to_payment                       92.20
payment_to_summary                        96.30
summary_to_order                          96.60

2. The conversion rate of mobile user is lower than that of desktop user in most segmentation, so we should take a look at out mobile app, and find how to improve it. 

In [41]:
CR_all_group 

BFYG_to_BS_CR  BS_to_payment_CR  \
Customer Type Delivery Type    Device Type                                    
Existing      Home Delivery    Desktop              95.05             96.20   
                               Mobile               94.75             92.30   
              Store Collection Desktop              92.75             93.60   
                               Mobile               91.45             91.55   
New           Home Delivery    Desktop              93.65             86.95   
                               Mobile               85.80             79.80   
              Store Collection Desktop              93.95             92.00   
                               Mobile               85.35             95.60   

                                            payment_to_summary_CR  \
Customer Type Delivery Type    Device Type                          
Existing      Home Delivery    Desktop                      95.90   
                               Mobile                       98.05   
              Store Collection Desktop                      97.25   
                               Mobile                       97.25   
New           Home Delivery    Desktop                      97.15   
                               Mobile                       92.95   
              Store Collection Desktop                      97.45   
                               Mobile                       93.85   

                                            summary_to_order_CR  
Customer Type Delivery Type    Device Type                       
Existing      Home Delivery    Desktop                    98.20  
                               Mobile                     98.75  
              Store Collection Desktop                    97.75  
                               Mobile                     96.70  
New           Home Delivery    Desktop                    97.05  
                               Mobile                     95.60  
              Store Collection Desktop                    94.80  
                               Mobile                     91.50

3. the performance of  recommendation engine could be improved, since the acceptance rate is only 36.5%, and the revenue per  added items is only 1.7

4. For the metric renevue per order(RPO), new customer has significantly low renevue. The RPO for new customer is 80 and the RPO for existing customer is 93. The performance of mobile user is also bad, the RPO for mobile user is 81 and the RPO for desktop user is 90. The RPO for store collection, which is 87, is also slightly lower than the RPO for home delivery, which is 91.

5. We have much less new user than existing user. the sum of orders come from new users is 120010 where the sum of orders come from existing user is 1237470 which is almost 10 time of orders of new users.

The metric I am using is convertion rate for each steps of checking out process. To measure the efficiency of recommendation engine I use recommendation acceptance rate for and revenue per customer and revenue per added item. 
I also used revenue per order, and looked of number of orders for each customer segmentation.

From the outcome from the analysis, I biggest problem is we don't have enough new users, so we should work harder in customer aquisition. Also there is several metrics indicated that mobile user has bad performance, so we should also look at our mobile app,and try to improve it. Lastly the performance of recommendation engine could be optimized. 